warehouse/internal/loadfiles/loadfiles.go

```diff
@@ -572,3 +587,105 @@ func toLoadFile(output LoadFileUpload, job *model.UploadJob) model.LoadFile {
 		UploadID:              &job.Upload.ID,
 	}
 }
+
+type fileWithMaxSize struct {
+	file    *model.StagingFile
+	maxSize int64
+}
+
+// GroupStagingFiles groups staging files based on their key characteristics
+// and then applies size constraints within each group. The maxSizeMB parameter controls the maximum size of any table within a group.
+func (lf *LoadFileGenerator) GroupStagingFiles(files []*model.StagingFile, maxSizeMB int) [][]*model.StagingFile {
+	// The fields in this struct are determined by the fields being used in the generateBaseRequest
+	type stagingFileGroupKey struct {
+		UseRudderStorage      bool
+		DestinationRevisionID string
+		TimeWindow            time.Time
+	}
+	groups := make(map[stagingFileGroupKey][]*fileWithMaxSize)
+
+	for _, file := range files {
+		key := stagingFileGroupKey{
+			UseRudderStorage:      file.UseRudderStorage,
+			DestinationRevisionID: file.DestinationRevisionID,
+			TimeWindow:            file.TimeWindow,
+		}
+		maxSize := int64(0)
+		for _, size := range file.BytesPerTable {
+			if size > maxSize {
+				maxSize = size
+			}
+		}
+		groups[key] = append(groups[key], &fileWithMaxSize{
+			file:    file,
+			maxSize: maxSize,
+		})
+	}
+
+	result := make([][]*model.StagingFile, 0, len(groups))
+
+	// For each group, apply size constraints
+	for _, group := range groups {
+		result = append(result, lf.groupBySize(group, maxSizeMB)...)
+	}
+	return result
+}
+
+// splits a group of staging files based on size constraints
+func (lf *LoadFileGenerator) groupBySize(files []*fileWithMaxSize, maxSizeMB int) [][]*model.StagingFile {
```

Comment:
offline discussion: 
1. `[]*fileWithMaxSize` has a file and largest table byte size in each file. Largest table byte size could be of `tracks` for one file and `identifies` for another. This won't give optimal batching in case if one file has more tracks, other file has less tracks. Ideally, we want large table size belonging to only one table per each file. 

2.  why move last item in slice to the current file index? Last file is going to have lesser table size 
```
                files[i] = files[len(files)-1]
				files = files[:len(files)-1]
```

Reply:
1. Fixed the code, to check for the only the table with max size across all files
2. Updated the element removal logic to avoid any changes in sorting order